File: apps/api/app/routes/exports.py
"""
Purpose: Expose export and evidence-pack API endpoints through the canonical
export workflow service.
Scope: Export triggering, listing, detail reads, evidence-pack assembly,
download, and idempotency-key preview.
Dependencies: FastAPI, request auth dependencies, export contracts, storage,
and the shared export service.
"""

from __future__ import annotations

from typing import Annotated
from uuid import UUID

from apps.api.app.dependencies.db import DatabaseSessionDependency
from apps.api.app.routes.auth import get_auth_service
from apps.api.app.routes.request_auth import AuthenticatedUserContext, RequestAuthDependency
from apps.api.app.routes.workflow_phase import require_active_close_run_phase
from fastapi import APIRouter, Depends, HTTPException, Query, Response, status
from services.auth.service import AuthService
from services.common.enums import ArtifactType, WorkflowPhase
from services.common.settings import AppSettings, get_settings
from services.contracts.export_models import (
    CreateExportRequest,
    DistributeExportRequest,
    EvidencePackBundle,
    ExportDetail,
    ExportListResponse,
    IdempotencyKeyResponse,
)
from services.contracts.storage_models import StorageBucketKind
from services.db.models.audit import AuditSourceSurface
from services.db.repositories.entity_repo import EntityUserRecord
from services.db.repositories.report_repo import ReportRepository
from services.exports.service import ExportService, ExportServiceError
from services.idempotency.service import build_idempotency_key
from services.storage.client import StorageClient

router = APIRouter(
    prefix="/entities/{entity_id}/close-runs/{close_run_id}/exports",
    tags=["exports"],
)

SettingsDependency = Annotated[AppSettings, Depends(get_settings)]
AuthServiceDependency = Annotated[AuthService, Depends(get_auth_service)]


def get_export_service(
    db_session: DatabaseSessionDependency,
) -> ExportService:
    """Construct the canonical export workflow service from request-scoped dependencies."""

    return ExportService(
        db_session=db_session,
        report_repository=ReportRepository(db_session=db_session),
    )


ExportServiceDependency = Annotated[ExportService, Depends(get_export_service)]


@router.post(
    "",
    response_model=ExportDetail,
    status_code=status.HTTP_201_CREATED,
    summary="Trigger a new export for a close run",
)
def trigger_export(
    entity_id: UUID,
    close_run_id: UUID,
    db_session: DatabaseSessionDependency,
    export_service: ExportServiceDependency,
    body: CreateExportRequest,
    auth_context: RequestAuthDependency,
) -> ExportDetail:
    """Trigger an export for one close run with idempotency protection."""

    require_active_close_run_phase(
        actor_user=_to_entity_user(auth_context),
        entity_id=entity_id,
        close_run_id=close_run_id,
        required_phase=WorkflowPhase.REVIEW_SIGNOFF,
        action_label="Export generation",
        db_session=db_session,
    )
    try:
        return export_service.trigger_export(
            actor_user=_to_entity_user(auth_context),
            entity_id=entity_id,
            close_run_id=close_run_id,
            request=body,
        )
    except ExportServiceError as error:
        _raise_export_http_error(error)


@router.get(
    "",
    response_model=ExportListResponse,
    summary="List exports for one close run",
)
def list_exports(
    entity_id: UUID,
    close_run_id: UUID,
    export_service: ExportServiceDependency,
    auth_context: RequestAuthDependency,
) -> ExportListResponse:
    """Return all export records for one close run in newest-first order."""

    try:
        return export_service.list_exports(
            actor_user=_to_entity_user(auth_context),
            entity_id=entity_id,
            close_run_id=close_run_id,
        )
    except ExportServiceError as error:
        _raise_export_http_error(error)


@router.get(
    "/{export_id}",
    response_model=ExportDetail,
    summary="Read one export detail",
)
def read_export_detail(
    entity_id: UUID,
    close_run_id: UUID,
    export_id: UUID,
    export_service: ExportServiceDependency,
    auth_context: RequestAuthDependency,
) -> ExportDetail:
    """Return one export record with full manifest and evidence-pack details."""

    try:
        return export_service.read_export_detail(
            actor_user=_to_entity_user(auth_context),
            entity_id=entity_id,
            close_run_id=close_run_id,
            export_id=export_id,
        )
    except ExportServiceError as error:
        _raise_export_http_error(error)


@router.post(
    "/{export_id}/distribute",
    response_model=ExportDetail,
    summary="Record management distribution for one export",
)
def distribute_export(
    entity_id: UUID,
    close_run_id: UUID,
    export_id: UUID,
    db_session: DatabaseSessionDependency,
    export_service: ExportServiceDependency,
    body: DistributeExportRequest,
    auth_context: RequestAuthDependency,
) -> ExportDetail:
    """Record one stakeholder release event for a completed export package."""

    require_active_close_run_phase(
        actor_user=_to_entity_user(auth_context),
        entity_id=entity_id,
        close_run_id=close_run_id,
        required_phase=WorkflowPhase.REVIEW_SIGNOFF,
        action_label="Management distribution",
        db_session=db_session,
    )
    try:
        return export_service.distribute_export(
            actor_user=_to_entity_user(auth_context),
            entity_id=entity_id,
            close_run_id=close_run_id,
            export_id=export_id,
            request=body,
            source_surface=AuditSourceSurface.DESKTOP,
            trace_id=None,
        )
    except ExportServiceError as error:
        _raise_export_http_error(error)


@router.post(
    "/evidence-pack",
    response_model=EvidencePackBundle,
    status_code=status.HTTP_201_CREATED,
    summary="Assemble and release an evidence pack for a close run",
)
def assemble_evidence_pack(
    entity_id: UUID,
    close_run_id: UUID,
    db_session: DatabaseSessionDependency,
    export_service: ExportServiceDependency,
    auth_context: RequestAuthDependency,
) -> EvidencePackBundle:
    """Assemble a downloadable evidence-pack bundle for one close run."""

    require_active_close_run_phase(
        actor_user=_to_entity_user(auth_context),
        entity_id=entity_id,
        close_run_id=close_run_id,
        required_phase=WorkflowPhase.REVIEW_SIGNOFF,
        action_label="Evidence-pack assembly",
        db_session=db_session,
    )
    try:
        return export_service.assemble_evidence_pack(
            actor_user=_to_entity_user(auth_context),
            entity_id=entity_id,
            close_run_id=close_run_id,
        )
    except ExportServiceError as error:
        _raise_export_http_error(error)


@router.get(
    "/evidence-pack",
    response_model=EvidencePackBundle,
    summary="Read the latest evidence pack for a close run",
)
def read_latest_evidence_pack(
    entity_id: UUID,
    close_run_id: UUID,
    export_service: ExportServiceDependency,
    auth_context: RequestAuthDependency,
) -> EvidencePackBundle:
    """Return the latest released evidence-pack bundle metadata for one close run."""

    try:
        evidence_pack = export_service.get_latest_evidence_pack(
            actor_user=_to_entity_user(auth_context),
            entity_id=entity_id,
            close_run_id=close_run_id,
        )
    except ExportServiceError as error:
        _raise_export_http_error(error)

    if evidence_pack is None or evidence_pack.storage_key is None:
        _raise_evidence_pack_not_found()
    return evidence_pack


@router.get(
    "/evidence-pack/download",
    summary="Download the latest evidence pack for a close run",
)
def download_evidence_pack(
    entity_id: UUID,
    close_run_id: UUID,
    settings: SettingsDependency,
    export_service: ExportServiceDependency,
    auth_context: RequestAuthDependency,
) -> Response:
    """Stream the latest released evidence-pack ZIP through the authenticated API surface."""

    try:
        evidence_pack = export_service.get_latest_evidence_pack(
            actor_user=_to_entity_user(auth_context),
            entity_id=entity_id,
            close_run_id=close_run_id,
        )
    except ExportServiceError as error:
        _raise_export_http_error(error)

    if evidence_pack is None or evidence_pack.storage_key is None:
        _raise_evidence_pack_not_found()

    payload = StorageClient.from_settings(settings).download_bytes(
        bucket_kind=StorageBucketKind.ARTIFACTS,
        object_key=evidence_pack.storage_key,
    )
    filename = f"evidence-pack-v{evidence_pack.version_no}.zip"
    return Response(
        content=payload,
        media_type="application/zip",
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )


@router.get(
    "/evidence-pack/idempotency-key",
    response_model=IdempotencyKeyResponse,
    summary="Preview the idempotency key for an evidence-pack action",
)
def preview_evidence_pack_idempotency_key(
    entity_id: UUID,
    close_run_id: UUID,
    export_service: ExportServiceDependency,
    auth_context: RequestAuthDependency,
    version: Annotated[int, Query(ge=1, description="Close-run version number.")] = 1,
) -> IdempotencyKeyResponse:
    """Return the deterministic idempotency key that would be used for an evidence-pack action."""

    try:
        export_service.list_exports(
            actor_user=_to_entity_user(auth_context),
            entity_id=entity_id,
            close_run_id=close_run_id,
        )
    except ExportServiceError as error:
        _raise_export_http_error(error)

    return IdempotencyKeyResponse(
        idempotency_key=build_idempotency_key(
            close_run_id=close_run_id,
            artifact_type=ArtifactType.EVIDENCE_PACK.value,
            action_qualifier="evidence_pack",
            version_override=version,
        ),
        close_run_id=str(close_run_id),
        artifact_type=ArtifactType.EVIDENCE_PACK.value,
    )


def _to_entity_user(session_result: AuthenticatedUserContext) -> EntityUserRecord:
    """Project the authenticated session user into the entity actor record."""

    return EntityUserRecord(
        id=session_result.user.id,
        email=session_result.user.email,
        full_name=session_result.user.full_name,
    )


def _raise_export_http_error(error: ExportServiceError) -> None:
    """Translate one export service error into the API's structured HTTP response."""

    raise HTTPException(
        status_code=error.status_code,
        detail={"code": str(error.code), "message": error.message},
    ) from error


def _raise_evidence_pack_not_found() -> None:
    """Raise the canonical not-found response for missing evidence-pack metadata."""

    raise HTTPException(
        status_code=404,
        detail={
            "code": "evidence_pack_not_found",
            "message": "No evidence pack has been assembled for this close run.",
        },
    )


__all__ = ["router"]

File: apps/desktop-ui/src/lib/exports.ts
/*
Purpose: Provide same-origin export and evidence-pack helpers for hosted close-run release pages.
Scope: Triggering exports, assembling evidence packs, and listing export records.
Dependencies: Browser Fetch APIs and existing `/api/entities/**` proxy routes.
*/

export type ExportSummary = {
  artifact_count: number;
  close_run_id: string;
  completed_at: string | null;
  created_at: string;
  distribution_count: number;
  failure_reason: string | null;
  id: string;
  idempotency_key: string;
  latest_distribution_at: string | null;
  status: string;
  version_no: number;
};

export type EvidencePackItem = {
  checksum: string | null;
  description: string | null;
  item_type: string;
  label: string;
  payload_ref: Record<string, unknown>;
  size_bytes: number | null;
  storage_key: string | null;
};

export type EvidencePackBundle = {
  checksum: string | null;
  close_run_id: string;
  generated_at: string;
  idempotency_key: string;
  items: readonly EvidencePackItem[];
  size_bytes: number | null;
  storage_key: string | null;
  version_no: number;
};

export type ExportArtifactEntry = {
  artifact_type: string;
  checksum: string;
  content_type: string;
  filename: string;
  idempotency_key: string;
  released_at: string;
  size_bytes: number;
  storage_key: string;
};

export type ExportManifest = {
  artifacts: readonly ExportArtifactEntry[];
  close_run_id: string;
  evidence_pack_ref: EvidencePackBundle | null;
  generated_at: string;
  version_no: number;
};

export type ExportDistributionRecord = {
  delivery_channel: string;
  distributed_at: string;
  distributed_by_user_id: string | null;
  id: string;
  note: string | null;
  recipient_email: string;
  recipient_name: string;
  recipient_role: string | null;
};

export type ExportDetail = ExportSummary & {
  distribution_records: readonly ExportDistributionRecord[];
  evidence_pack: EvidencePackBundle | null;
  manifest: ExportManifest | null;
};

export type DistributeExportRequest = {
  delivery_channel: string;
  note?: string | null;
  recipient_email: string;
  recipient_name: string;
  recipient_role?: string | null;
};

export const EXPORT_DELIVERY_CHANNELS = [
  "secure_email",
  "management_portal",
  "board_pack",
  "file_share",
] as const;

export class ExportApiError extends Error {
  constructor(
    message: string,
    public readonly status: number,
    public readonly code: string | null = null,
  ) {
    super(message);
    this.name = "ExportApiError";
  }
}

function buildEntityProxyPath(entityId: string, pathSegments: readonly string[]): string {
  const encodedSegments = [entityId, ...pathSegments].map((segment) => encodeURIComponent(segment));
  return `/api/entities/${encodedSegments.join("/")}`;
}

async function requestJson<T>(url: string, init: RequestInit = {}): Promise<T> {
  const response = await fetch(url, {
    ...init,
    cache: "no-store",
    credentials: "include",
    headers: {
      Accept: "application/json",
      "Content-Type": "application/json",
      ...init.headers,
    },
  });
  const payload = await parseJsonResponse(response);
  if (!response.ok) {
    const detail = isRecord(payload) ? payload.detail : null;
    throw new ExportApiError(
      isRecord(detail) && typeof detail.message === "string"
        ? detail.message
        : `Request failed with status ${response.status}.`,
      response.status,
      isRecord(detail) && typeof detail.code === "string" ? detail.code : null,
    );
  }
  return payload as T;
}

export async function listExports(
  entityId: string,
  closeRunId: string,
): Promise<readonly ExportSummary[]> {
  const payload = await requestJson<{ exports: ExportSummary[] }>(
    buildEntityProxyPath(entityId, ["close-runs", closeRunId, "exports"]),
    { method: "GET" },
  );
  return payload.exports;
}

export async function triggerExport(entityId: string, closeRunId: string): Promise<ExportDetail> {
  return requestJson<ExportDetail>(
    buildEntityProxyPath(entityId, ["close-runs", closeRunId, "exports"]),
    {
      method: "POST",
      body: JSON.stringify({ action_qualifier: "full_export" }),
    },
  );
}

export async function readExportDetail(
  entityId: string,
  closeRunId: string,
  exportId: string,
): Promise<ExportDetail> {
  return requestJson<ExportDetail>(
    buildEntityProxyPath(entityId, ["close-runs", closeRunId, "exports", exportId]),
    { method: "GET" },
  );
}

export async function assembleEvidencePack(
  entityId: string,
  closeRunId: string,
): Promise<EvidencePackBundle> {
  return requestJson<EvidencePackBundle>(
    buildEntityProxyPath(entityId, ["close-runs", closeRunId, "exports", "evidence-pack"]),
    { method: "POST", body: JSON.stringify({}) },
  );
}

export async function distributeExport(
  entityId: string,
  closeRunId: string,
  exportId: string,
  payload: DistributeExportRequest,
): Promise<ExportDetail> {
  return requestJson<ExportDetail>(
    buildEntityProxyPath(entityId, ["close-runs", closeRunId, "exports", exportId, "distribute"]),
    {
      method: "POST",
      body: JSON.stringify(payload),
    },
  );
}

export async function readLatestEvidencePack(
  entityId: string,
  closeRunId: string,
): Promise<EvidencePackBundle | null> {
  try {
    return await requestJson<EvidencePackBundle>(
      buildEntityProxyPath(entityId, ["close-runs", closeRunId, "exports", "evidence-pack"]),
      { method: "GET" },
    );
  } catch (error) {
    if (
      error instanceof ExportApiError &&
      error.status === 404 &&
      error.code === "evidence_pack_not_found"
    ) {
      return null;
    }
    throw error;
  }
}

export function buildEvidencePackDownloadPath(entityId: string, closeRunId: string): string {
  return buildEntityProxyPath(entityId, ["close-runs", closeRunId, "exports", "evidence-pack", "download"]);
}

async function parseJsonResponse(response: Response): Promise<unknown> {
  const text = await response.text();
  if (text.length === 0) {
    return null;
  }

  try {
    return JSON.parse(text) as unknown;
  } catch {
    return null;
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null;
}
